Examples/c++/Linked List/DLList.h

```diff
@@ -0,0 +1,401 @@
+			m_SentinelNode->Previous = m_SentinelNode;
+		}
+
+		~DLList()
+		{
+			Clear();
+			delete m_SentinelNode;
+		}
+
+		void PushFront(const T& data)
+		{
+			InsertAfter(m_SentinelNode, data);
+		}
+
+		void PushFront(T&& data)
+		{
+			InsertAfter(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceFront(Args&&... args)
+		{
+			return EmplaceAfter(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void PushBack(const T& data)
+		{
+			InsertBefore(m_SentinelNode, data);
+		}
+
+		void PushBack(T&& data)
+		{
+			InsertBefore(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceBack(Args&&... args)
+		{
+			return EmplaceBefore(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void Insert(Iterator where, const T& data)
+		{
+			InsertBefore((NodePtr)where, data);
+		}
+
+		void Insert(Iterator where, T&& data)
+		{
+			InsertBefore((NodePtr)where, std::move(data));
+		}
+
+		template<typename... Args>
+		T& Emplace(Iterator where, Args&&... args)
+		{
+			return EmplaceBefore((NodePtr)where, std::forward<Args>(args)...);
+		}
+
+		void PopFront()
+		{
+			DeleteNode(m_SentinelNode->Next);
+		}
+
+		void PopBack()
+		{
+			ASSERT(!Empty(), "PopBack() called on an empty list!");
+			DeleteNode(m_SentinelNode->Previous);
+		}
+
+		void Erase(Iterator where)
+		{
+			ASSERT((NodePtr)where != m_SentinelNode, "Invalid Iterator");
+			DeleteNode(where);
+		}
+
+		void MoveToBack(Iterator from, Iterator to)
+		{
+			if (m_Size <= 1 || to == end())
+				return;
+
+			NodePtr startNode = (NodePtr)from;
+			NodePtr endNode = ((NodePtr)to)->Previous;
+
+			NodePtr startNodePrevious = startNode->Previous;
+
+			endNode->Next = m_SentinelNode;
+			m_SentinelNode->Previous->Next = startNode;
+			startNode->Previous = m_SentinelNode->Previous;
+
+			m_SentinelNode->Previous = endNode;
+			startNodePrevious->Next = (NodePtr)to;
+			((NodePtr)to)->Previous = startNodePrevious;
+
+		}
+
+		void Reverse()
+		{
+			NodePtr current = m_SentinelNode->Next;
+
+			while (current != m_SentinelNode)
+			{
+				std::swap(current->Next, current->Previous);
+				current = current->Previous;
+			}
+
+			std::swap(m_SentinelNode->Next, m_SentinelNode->Previous);
+			
+			m_SentinelNode->Previous->Next = m_SentinelNode;
+			m_SentinelNode->Next->Previous = m_SentinelNode;
+		}
+
+		void Reverse(Iterator from, Iterator to)
+		{
+			if (from == to)
+				return;
+
+			NodePtr beginNode = (NodePtr)from;
+			NodePtr endNode = ((NodePtr)to)->Previous;
+			NodePtr beginNodePrevious = beginNode->Previous;
+			NodePtr endNodeNext = (NodePtr)to;
+
+			endNode->Next = nullptr;
+
+			NodePtr current = beginNode;
+			NodePtr prev = nullptr;
+
+			while (current != nullptr)
+			{
+				std::swap(current->Next, current->Previous);
+				prev = current;
+				current = current->Previous;
+			}
+
+			beginNodePrevious->Next = prev;
+			beginNode->Next = endNodeNext;
+		}
+
+		void Clear()
```

Comment:
Do we need this function public?

Reply:
This works the same as the STL clear functions. I left it public so if the user wants to clear the whole list they have that option.